Added support for analyzing sources using PMD

ajaychandran opened the pull request:
This PR adds support for analyzing sources using [PMD](https://docs.pmd-code.org/latest/index.html) via the `mill.javalib.PmdModule`.

- Custom rules/rulesets are supported.
- Companion external module included for global usage.
- Offline mode is supported.

All [PMD languages](https://docs.pmd-code.org/latest/tag_PmdCapableLanguage.html) are supported, theoretically, but test coverage is limited to Java.

- There are only 2 rules available for [Kotlin](https://docs.pmd-code.org/latest/pmd_rules_kotlin.html).
- There are no rules for [Scala](https://docs.pmd-code.org/latest/pmd_rules_scala.html).

Accordingly, the API has been marked experimental for now.

Resolves #5379.

ajaychandran committed:
Added support for PMD (Programming Mistake Detector)

ajaychandran committed:
Addressed review comments.
- Moved dependency definition to Deps.RuntimeDeps.
- Use PathRef values for pmdRulesets and pmdInputs.
- Added format argument to pmd command to replace pmdFormat.

ajaychandran commented on libs/scalalib/src/mill/javalib/pmd/PmdModule.scala:
```diff
@@ -0,0 +1,206 @@
+   *  - ''workspace'', otherwise
+   * @note Path values can be specified relative to [[moduleDir]].
+   */
+  def pmdRulesets: Task[Seq[String]] = Task {
```
Updated both `pmdRulesets` and `pmdInputs`.

The former is a list of configuration files. The default value was selected to handle the two use-cases where rulesets are managed on a module basis or globally in the workspace module. It is not safe to use files at different levels, we don't know what the user intends in these cases.
 
The latter is the list of a combination of source files and/or folder roots with source files to be analyzed. In practice, this will always be `JavaModule.sources`.

I will add the  [--exclude <filepath+>](https://docs.pmd-code.org/latest/pmd_userdocs_cli_reference.html#file-collection-options) option for skipping paths.

lefou replied:
This should be a `Task.Sources` to capture changes. This will likely eliminate the need for the `withFilesystemCheckerDisabled` tweak, too.

I don't get why this might end up with `JavaModule.sources` though, since it typically starts with `src` and not `ruleset` and you don't search recursively.

ajaychandran committed:
- Added support for --exclude options
- Use -d options in place of positional values
- Remove redundant option

ajaychandran commented on libs/scalalib/src/mill/javalib/pmd/PmdModule.scala:
```diff
@@ -0,0 +1,206 @@
+   * Output [[https://docs.pmd-code.org/latest/pmd_userdocs_report_formats.html format]] of the
+   * analysis report. Defaults to ''text''.
+   */
+  def pmdFormat: Task[String] = Task("text")
```
The positional values for inputs were replaced with an explicit option (`-d`). This eliminated the quirk and the `format` option.

lefou commented on libs/scalalib/src/mill/javalib/pmd/PmdModule.scala:
```diff
@@ -0,0 +1,196 @@
+  /**
+   * File to which output report is written. If empty, the report is rendered to standard output.
+   */
+  def pmdReportFile: Task[Option[PathRef]] = Task(Option.empty[PathRef])
```
What is this task for? If the `pmd` command needs to write to a file, it should be located under `pmd`'s dest folder.

ajaychandran replied:
This has been removed.

lefou commented on libs/scalalib/src/mill/javalib/pmd/PmdModule.scala:
```diff
@@ -0,0 +1,196 @@
+   */
+  def pmdInputs: Task[Seq[PathRef]] = this match
+    case self: JavaModule => Task(self.sources())
+    case _ => Task {
+        BuildCtx.withFilesystemCheckerDisabled {
+          os.list.stream(moduleDir)
+            .collect:
+              case path
+                  if os.isDir(path) && path.relativeTo(moduleDir).segments.head.startsWith("src") =>
+                PathRef(path)
+            .toSeq
+        }
+      }
```
This should be a `Task.Sources` to account for changed files.

ajaychandran replied:
Done.

ajaychandran committed:
Use Task.Sources for pmdRulesets and pmdInputs.
Removed pmdReportFile.

ajaychandran committed:
- Use Task.Input when reading options from files
- Support root module usage
- Added test for --use-version

ajaychandran committed:
Fixes post rebase

ajaychandran committed:
Cleanup

ajaychandran committed:
Cleaned and rebased.

ajaychandran commented:
This is mostly ready.

The idiomatic way to support [third party libraries](https://docs.pmd-code.org/latest/pmd_userdocs_3rdpartyrulesets.html) would be to add a member (`pmdMvnDeps`) for corresponding Maven dependencies. I will work on this over the weekend.

ajaychandran committed:
Added support for third party rulesets

ajaychandran committed:
Merge remote-tracking branch 'upstream/main' into pmd

# Conflicts:
#	example/javalib/linting/6-pmd/build.mill
#	libs/jvmlib/package.mill
#	mill-build/src/millbuild/Deps.scala

ajaychandran committed:
Replaced existing implmentation for PMD support

ajaychandran committed:
Added example for defining and using custom rules

ajaychandran commented:
Added support for defining and using custom rules/rulesets with `pmdMvnDeps` and `pmdModuleDeps`. Added an example for the latter.

The existing implementation from #5393 was removed and the PR description was updated.

ajaychandran commented on libs/jvmlib/src/mill/javalib/pmd/PmdModule.scala:
```diff
@@ -1,175 +1,165 @@
+   *       can be used after adding the dependencies here.
+   */
+  def pmdMvnDeps: Task[Seq[Dep]] = this match
+    case _: BaseModule => Task(Seq(mvn"$pmdDist"))
+    case _: JavaModule => Task(Seq(mvn"$pmdCli", mvn"$pmdJava"))
+    case _ => Task(Seq(mvn"$pmdCli"))
```
Or this could be simplified to always use `pmdDist`.

ajaychandran commented:
Mima check is failing. Should I add exceptions?
```
[1496] Found 2 issue when checking against com.lihaoyi:mill-libs-jvmlib-api_3:VersionConstraint.Lazy(1.0.0-RC3-10-2e07f5, VersionInterval(None, None, false, false), Some(Version(1.0.0-RC3-10-2e07f5)), None)
[1496]  * static method pmdVersion()java.lang.String in class mill.jvmlib.api.Versions does not have a correspondent in current version
[1496]    filter with: ProblemFilter.exclude[DirectMissingMethodProblem]("mill.jvmlib.api.Versions.pmdVersion")
[1496]  * method pmdVersion()java.lang.String in object mill.jvmlib.api.Versions does not have a correspondent in current version
[1496]    filter with: ProblemFilter.exclude[DirectMissingMethodProblem]("mill.jvmlib.api.Versions.pmdVersion")
[1473] Binary compatibility check passed
[1473/1656, 1 failed] ======================= __.mimaReportBinaryIssues ====================== 21s
1 tasks failed
libs.jvmlib.api.mimaReportBinaryIssues Failed binary compatibility check! Found 2 potential problems
```

lihaoyi commented:
@ajaychandran yes just add an exclusion for now

ajaychandran committed:
Added MiMa exclusions (for replacing changes from #5393)

ajaychandran commented:
CI failure seems unrelated.
```
[7065]     File "/home/runner/.cache/nce/ba7257887eb1da08c71c410700e65849b7716ee6237f3e86558d33802728b3ec/bindings/shiv_root/science.pyz_ee33d73d113bc985a5864e6f8cba0864a510cb1e41b1ee0166167c2dc244968c/site-packages/science/providers/python_build_standalone.py", line 204, in create
[7065]       release = release_data["name"]
[7065]                 ~~~~~~~~~~~~^^^^^^^^
[7065]   builtins.KeyError: 'name'
[7065]   1 tasks failed
[7065]   show 1 tasks failed
[7065]   qux.bundle os.SubprocessException: Result of /home/runner/work/mill/mill/out/example/extending/python/4-python-libs-bundle/local/daemon/testForked.dest/sandbox/run-2/out/qux/pythonExe.dest/venv/bin/python3?: 1
[7065]   
[7065]     os.SubprocessException$.apply(Model.scala:211)
[7065]       os.proc.call(ProcessOps.scala:232)
[7065]       os.call$.apply(ProcessOps.scala:42)
[7065]       build_.package_$PythonModule.bundle$$anonfun$1$$anonfun$2(build.mill:86)
```

ajaychandran commented:
Apologies. I did misread the intention.
I will submit another PR with the suggested improvements.

> I think this PR adds valuable features, which we might want to merge, like out-of-the-box support for per-module configurations, language specific versions, PMD runner options and third-party rulesets.
